### PR TITLE
fix: VLM packed sequence padding and SP scatter ordering for CP+SP

### PR DESCRIPTION
**Pull request by cuichenx**
Three fixes for VLM training with context parallelism (CP) + sequence parallelism (SP) + packed sequences (THD format):

1. vlm_step.py: Compute pad_to_multiple_of as cp_size * tp_size when SP is enabled, instead of cp_size * 2. After CP splitting, each rank gets total_packed_len / cp_size tokens, which must be divisible by tp_size for SP's reduce_scatter. The minimum correct padding is lcm(2*cp_size, cp_size*tp_size) = cp_size*tp_size (when tp_size>=2). Reference: megatron/core/models/multimodal/context_parallel.py

2. modeling_ministral3.py, modeling_gemma3_vl.py: Apply SP scatter AFTER CP slice, not before. When VLM models pass decoder_input directly to the language model, the embedding layer (which normally handles SP scatter) is bypassed. The correct order is CP slice first, then SP scatter, matching Megatron Core's LLaVA pattern (llava_model.py:747).

3. vlm_step.py: Detect collator pad token from tokens[i, -1] instead of hardcoding pad_token_id=0. VLM collators use model-specific pad tokens (e.g., token 11 for Ministral3), so the old logic could not strip collator padding, causing all packed sequences to appear equal-length.

Verified with 20 iterations of Ministral3-3B on CORD-v2, TP=4, CP=2, SP=True, MBS=3, packed sequences on 8x H100.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

# Changelog

- Add specific line by line info of high level changes in this PR.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed sequence-parallel scatter timing in vision-language models to execute after context-parallel slicing for correct distributed training behavior.
  * Improved padding token detection in batch packing to correctly handle non-zero padding tokens from the collator.

* **Performance**
  * Optimized batch padding calculations for context-parallel and sequence-parallel configurations.

### CODE REVIEW

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2932/commits/a833d40f4ff36a1a908bff9464a48e99441c86cf

**Review by yaoyu-33 — DISMISSED** [review]

File: src/megatron/bridge/training/vlm_step.py [hunk 1]
```diff
@@ -94,6 +95,13 @@ def get_batch_from_iterator(
         else:
             _batch_required_keys[key] = None
 
+    # Preserve collator's 2D padding mask for sequence packing length detection.
+    # skip_getting_attention_mask_from_dataset may discard the attention_mask for model
+    # forward, but packing still needs it to identify real vs padding positions.
+    raw_attn = batch.get("attention_mask")
+    if isinstance(raw_attn, torch.Tensor) and raw_attn.dim() == 2:
+        _batch_required_keys["_padding_mask"] = raw_attn.cuda(non_blocking=True)
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/vlm_step.py:103]: what padding mask do?

Comment (cuichenx) [reply src/megatron/bridge/training/vlm_step.py:103]: there was a bug before that packs the padded sequences (unnecessary padding in the pack). this removes it. see usage below

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2932/commits/ad5b6c9661d277ea24a7b215932ad765e8274396

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2932/commits/06e6fab7749eb00a68c2877f67329894ec609e1f

**Review by yaoyu-33 — APPROVED** [review]

**Comment by pruprakash** [comment]
## QA RCCA Analysis

### 1. Fix Reference
- **PR**: #2932 - fix: VLM packed sequence padding and SP scatter ordering for CP+SP
- **Issue**: PR is the fix itself (no linked issue)

### 2. Root Cause
Three fixes for VLM training with CP + SP + packed sequences (THD format):
1. Incorrect `pad_to_multiple_of` calculation (should be `cp_size * tp_size`, not `cp_size * 2`)
2. SP scatter applied in wrong order in Ministral3 and Gemma3-VL
3. After CP splitting, each rank needs tokens divisible by tp_size for SP reduce_scatter

### 3. Trigger Configuration
- VLM training with context parallelism (CP)
- Sequence parallelism (SP) enabled
- Packed sequences (THD format)

### 4. Nature of the Bug
**Classification**: CODE BUG - Incorrect padding and scatter ordering for CP+SP VLM training

### 5. Existing Test Coverage
**In Fix PR**: YES - 1 test file:
- `tests/unit_tests/training/test_vlm_step.py`

**In NMFW Tests**: Not specifically covering CP+SP packed sequence padding

### 6. Coverage Assessment
| Test Type | Exists | Covers Bug |
|-----------|--------|------------|
| Fix PR unit tests | YES | YES |
| NMFW regression tests | NO | N/A |

### 7. New Regression Test
**NOT NEEDED** - Fix PR includes unit tests for VLM step with CP+SP.

### 8. Conclusion
**Verdict**: ADEQUATE COVERAGE - Fix PR includes unit tests for VLM packed sequence handling.
